app/models/go_game.rb

```diff
@@ -0,0 +1,143 @@
+# this class represents everything needed to build a square Go board,
+# track the state of the game, and possibly produce a game record based
+# on the stones placed on the board.
+class GoGame
```

Comment:
Wow!  There is so much going on here.  Congrats on making it all work.

Despite remembering your talk, I am not very familiar with the rules of Go, but I tested capturing, next player moves, and legal and illegal moves.

Your game handles them all really well! 🚀 

I did find some of this code hard to follow, specifically the code updating the groups.  

But I _love_ the Structs in here helping you save intermediate state.
I really like all the comments.
You've named things particularly well.

If you wanted to refactor, I might consider:
* making the structs their own file
* creating a service object to handle the logic inside the update_groups piece

Reply:
i will work on cleaning it up more when i get time. i really appreciate you taking time to do a review @Gmfholley thank you so much!